src/metabase/driver/druid.clj

```diff
@@ -57,15 +58,39 @@
 (defn- do-query [details query]
   {:pre [(map? query)]}
   (ssh/with-ssh-tunnel [details-with-tunnel details]
-    (try (vec (POST (details->url details-with-tunnel "/druid/v2"), :body query))
-         (catch Throwable e
-           ;; try to extract the error
-           (let [message (or (u/ignore-exceptions
-                               (:error (json/parse-string (:body (:object (ex-data e))) keyword)))
-                             (.getMessage e))]
-             (log/error (u/format-color 'red "Error running query:\n%s" message))
-             ;; Re-throw a new exception with `message` set to the extracted message
-             (throw (Exception. message e)))))))
+    (try
+      (POST (details->url details-with-tunnel "/druid/v2"), :body query)
+      (catch Throwable e
+        ;; try to extract the error
+        (let [message (or (u/ignore-exceptions
+                            (:error (json/parse-string (:body (:object (ex-data e))) keyword)))
+                          (.getMessage e))]
+          (log/error (u/format-color 'red "Error running query:\n%s" message))
+          ;; Re-throw a new exception with `message` set to the extracted message
+          (throw (Exception. message e)))))))
+
+(defn- do-query-with-cancellation [details query]
+  (let [query-id  (get-in query [:context :queryId])
+        query-fut (future (do-query details query))]
+    (try
+      ;; Run the query in a future so that this thread will be interrupted, not the thread running the query (which is
+      ;; not interrupt aware)
+      @query-fut
+      (catch InterruptedException interrupted-ex
+        ;; The future has been cancelled, if we ahve a query id, try to cancel the query
+        (if-not query-id
+          (log/warn interrupted-ex "Client closed connection, no queryId found, can't cancel query")
+          (ssh/with-ssh-tunnel [details-with-tunnel details]
+            (log/warnf "Client closed connection, cancelling Druid queryId '%s'" query-id)
+            (try
+              ;; If we can't cancel the query, we don't want to hide the original exception, attempt to cancel, but if
+              ;; we can't, we should rethrow the InterruptedException, not an exception from the cancellation
+              (DELETE (details->url details-with-tunnel (format "/druid/v2/%s" query-id)))
+              ;; Propogate the exception, will cause any other catch/finally clauses to fire
+              (throw interrupted-ex)
+              (catch Exception cancel-ex
+                (log/warnf cancel-ex "Failed to cancel Druid query with queryId" query-id)
+                (throw interrupted-ex)))))))))
```

Comment:
Wouldn’t a finally be cleaner/clearer?

Reply:
I agree, remove line 90 and 93, swap it for a finally that throws :+1: